src/lib.rs

```diff
@@ -1574,14 +1620,150 @@ mod test {
+        tester.clear();
+    }
+
     #[test]
     #[cfg(feature = "std")]
     fn test_show() {
-        use std::string::ToString;
-        assert_eq!(format!("{}", _2), "2".to_string());
-        assert_eq!(format!("{}", _1_2), "1/2".to_string());
-        assert_eq!(format!("{}", _0), "0".to_string());
-        assert_eq!(format!("{}", Ratio::from_integer(-2)), "-2".to_string());
+        // Test:
+        // :b :o :x, :X, :?
+        // alternate or not (#)
+        // positive and negative
+        // padding
+        // does not test precision (i.e. truncation)
+        assert_eq!(&format!("{}", _2), "2");
+        assert_eq!(&format!("{}", _1_2), "1/2");
+        assert_eq!(&format!("{}", -_1_2), "-1/2"); // test negatives
+        assert_eq!(&format!("{}", _0), "0");
+        assert_eq!(&format!("{}", -_2), "-2");
+        assert_eq!(&format!("{:b}", _2), "10");
+        assert_eq!(&format!("{:b}", _1_2), "1/10");
+        assert_eq!(&format!("{:b}", _0), "0");
+        assert_eq!(&format!("{:#b}", _1_2), "0b1/0b10");
+        assert_eq!(&format!("{:010b}", _1_2), "0000001/10");
+        assert_eq!(&format!("{:#010b}", _1_2), "0b001/0b10");
+        let half_i8: Ratio<i8> = Ratio::new(1_i8, 2_i8);
+        assert_eq!(&format!("{:b}", -half_i8), "11111111/10");
+        assert_eq!(&format!("{:#b}", -half_i8), "0b11111111/0b10");
+
+        assert_eq!(&format!("{:o}", _8), "10");
+        assert_eq!(&format!("{:o}", _1_8), "1/10");
+        assert_eq!(&format!("{:o}", _0), "0");
+        assert_eq!(&format!("{:#o}", _1_8), "0o1/0o10");
+        assert_eq!(&format!("{:010o}", _1_8), "0000001/10");
+        assert_eq!(&format!("{:#010o}", _1_8), "0o001/0o10");
+        assert_eq!(&format!("{:o}", -half_i8), "377/2");
+        assert_eq!(&format!("{:#o}", -half_i8), "0o377/0o2");
+
+        assert_eq!(&format!("{:x}", _16), "10");
+        assert_eq!(&format!("{:x}", _15), "f");
+        assert_eq!(&format!("{:x}", _1_16), "1/10");
+        assert_eq!(&format!("{:x}", _1_15), "1/f");
+        assert_eq!(&format!("{:x}", _0), "0");
+        assert_eq!(&format!("{:#x}", _1_16), "0x1/0x10");
+        assert_eq!(&format!("{:010x}", _1_16), "0000001/10");
+        assert_eq!(&format!("{:#010x}", _1_16), "0x001/0x10");
+        assert_eq!(&format!("{:x}", -half_i8), "ff/2");
+        assert_eq!(&format!("{:#x}", -half_i8), "0xff/0x2");
+
+        assert_eq!(&format!("{:X}", _16), "10");
+        assert_eq!(&format!("{:X}", _15), "F");
+        assert_eq!(&format!("{:X}", _1_16), "1/10");
+        assert_eq!(&format!("{:X}", _1_15), "1/F");
+        assert_eq!(&format!("{:X}", _0), "0");
+        assert_eq!(&format!("{:#X}", _1_16), "0x1/0x10");
+        assert_eq!(&format!("{:010X}", _1_16), "0000001/10");
+        assert_eq!(&format!("{:#010X}", _1_16), "0x001/0x10");
+        assert_eq!(&format!("{:X}", -half_i8), "FF/2");
+        assert_eq!(&format!("{:#X}", -half_i8), "0xFF/0x2");
+
+        let _one_tehth_1_f = Ratio {
```

Comment:
Nit:

```suggestion
        let _one_tenth_1_f = Ratio {
```

However, it's pretty weird to be testing `Ratio<f32>` at all -- do you have a specific motivation for that?

Reply:
I believe I wanted to make sure that `Ratio`s properly implement `UpperExp` and `LowerExp` for types that implement those traits. The only such types available in the language are floats. I actually just merged a PR in nightly that will add *Exp formatting to integers, but I wrote this before that (and we need to support Rust 1.31).

Reply:
I found rust-lang/rust#66721, congrats!

We could add back a build script and use `autocfg` to probe for available support, giving you a `cfg` flag to guard the test case. It will only work on nightlies for now, but that will trickle down.